Desired nodes API

fcofdez opened the pull request:
This commit adds the Desired Nodes API, allowing orchestrators
that manage Elasticsearch clusters to let the system know about the 
current/planned topology that the cluster will run on. 
This allows the system to take better decisions based on the entire 
cluster topology, including nodes that will be added/removed in the 
near future.

This commit adds the basic endpoints to manage the desired nodes
state:
- GET /_internal/desired_nodes
- PUT /_internal/desired_nodes/<history_id>/<version>
- DELETE /_internal/desired_nodes

fcofdez committed:
Add boilerplate

fcofdez committed:
More boilerplate

fcofdez committed:
More progress

fcofdez committed:
Bind DesiredNodesService

fcofdez committed:
Add tests for serialization

fcofdez committed:
Minor adjustments

fcofdez committed:
More progress

fcofdez committed:
Remove unused code

fcofdez committed:
Add more tests and validations

fcofdez committed:
More tests

fcofdez committed:
Progress

fcofdez committed:
Add more testing

fcofdez committed:
Move tests around

fcofdez committed:
Progress

fcofdez committed:
Fix mistake

fcofdez committed:
Wire everything

fcofdez committed:
Add get and delete methods

fcofdez committed:
rest spec, yaml tests and some other tests

fcofdez committed:
Add more yaml tests

fcofdez committed:
More tests

fcofdez committed:
Cleanup

fcofdez committed:
Remove unused import

fcofdez committed:
Fix tests

fcofdez committed:
Add docs

fcofdez committed:
Update docs

fcofdez committed:
Fix test

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

fcofdez commented:
I've to fix a couple of things around docs and operators permissions, but I think this is ready for a first round of reviews.

fcofdez committed:
Skip yaml tests in previous versions

fcofdez committed:
Fix invalid doc ref

fcofdez committed:
Add desired nodes actions to operator registry

fcofdez committed:
Fix docs test

fcofdez committed:
Update docs/changelog/82975.yaml

elasticsearchmachine commented:
Hi @fcofdez, I've created a changelog YAML for you.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:
```diff
@@ -0,0 +1,115 @@
+            }
+        }
+
+        if (unknownSettings.isEmpty() == false && node.version().onOrBefore(Version.CURRENT)) {
```
I think we would want something like the version barrier here, i.e., not allow going backwards in node versions. Can certainly wait for a follow-up.

henningandersen commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -37,13 +37,11 @@
-        Runtime.getRuntime().availableProcessors(),
-        Property.NodeScope
-    );
+    public static final Setting<Integer> NODE_PROCESSORS_SETTING = createNodeProcessorsSetting(Runtime.getRuntime().availableProcessors());
```
Can we instead add a new unregistered setting with the available processors? And use that as the default value for the `node.processors` value. Then the validation could simply override the new setting in the `Settings` object.

fcofdez replied:
There's a problem with that approach, if the setting is invalid it will report that the "hidden" setting is invalid and that might be confusing.

henningandersen replied:
I'd consider that a bug, since we should validate the input we get to desired nodes - and that should then also be valid on the hidden setting. Perhaps we can validate each internal setting in isolation and report an API level error on the invalid `desired_nodes` parameter (like `processors`)?

fcofdez replied:
I'm not sure I understand what you mean. Currently we validate the `node.processors` setting using the provided desired node `processors`. Are you suggesting that we shouldn't override this setting and instead validate it without using the regular settings validation path?

henningandersen replied:
We spoke about this offline. I am ok to defer this to a follow-up, the situations where this matters is not that important, but we should address it eventually. We can ensure the validation works by validating the `UpdateDesiredNodesRequest` to ensure it does not contain any invalid specifications (`processors<=0` for instance).

henningandersen reviewed (commented):
Thanks Francisco, this mostly looks good. I did not get through it all yet but wanted to share the comments I have early.

sethmlarson commented on docs/reference/cluster/get-desired-nodes.asciidoc:
```diff
@@ -0,0 +1,102 @@
+[source,console-result]
+--------------------------------------------------
+{
+    "desired_nodes": {
```
Also there seems to be a name clash between `version` and `nodes.version`, is there anything we can rename the top-level `version` to?

fcofdez replied:
We can rename it to `history_version`?

henningandersen replied:
I am not sure the name clash is really an issue? But we can rename to "desired_nodes_version" and "desired_nodes_history_d" to make it explicit? We can also rename "nodes.version" to "nodes.node_version".

sethmlarson commented on docs/reference/cluster/update-desired-nodes.asciidoc:
```diff
@@ -0,0 +1,112 @@
+[source,console-result]
+--------------------------------------------------
+{
+  "acknowledged": true
```
Many APIs have `acknowledged: true` even if they can fail and return an error so leaving as-is would be fine here.

sethmlarson reviewed (changes requested):
Few comments for you! Also general comment, the namespace for this API is currently `desired_nodes` but has a path prefix of `/_cluster/` but appears to be at least partially-related to autoscaling? We try to reserve creating new namespaces for brand new sets of features that don't fit into existing namespace. At a minimum this feature appears to fit into `autoscaling` or `cluster`.

What are thoughts on switching to `autoscaling.get_desired_nodes`, etc for the JSON specs and using `/_autoscaling/` as a path prefix?

fcofdez commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:
```diff
@@ -0,0 +1,364 @@
+            - { settings: { "node.external_id": "instance-000187", "http.tcp.keep_idle": -1000 }, processors: 8, memory: "64gb", storage: "128gb", version: "7.16.0" }
+  - match: { status: 400 }
+  - match: { error.type: illegal_argument_exception }
+  - match: { error.reason: "Nodes in positions [0] contain invalid settings" }
```
One of the validations include checking that `external_id` is present, that's why I decided to go with the list position. Maybe we can include it if it's present? Something like `Nodes in positions [(0, <external_id>)] contain invalid settings`?

fcofdez commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:
```diff
@@ -0,0 +1,132 @@
+                return currentState;
+            }
+
+            if (currentDesiredNodes.hasSameVersion(proposedDesiredNodes) && currentDesiredNodes.equals(proposedDesiredNodes) == false) {
```
Right 🤦

fcofdez committed:
First round of reviews

fcofdez committed:
Merge branch 'desired-nodes-api' of github.com:fcofdez/elasticsearch into desired-nodes-api

fcofdez committed:
Fix changelog

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

elasticsearchmachine commented:
Hi @fcofdez, I've updated the changelog YAML for you.

fcofdez committed:
Update docs/changelog/82975.yaml

elasticsearchmachine commented:
Hi @fcofdez, I've updated the changelog YAML for you.

fcofdez committed:
Update docs/changelog/82975.yaml

fcofdez committed:
Fix docs

fcofdez committed:
Merge branch 'desired-nodes-api' of github.com:fcofdez/elasticsearch into desired-nodes-api

fcofdez commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/cluster.desired_nodes/10_basic.yml:
```diff
@@ -0,0 +1,364 @@
+        version: 1
+        body:
+            nodes:
+              - { settings: { "node.external_id": "instance-000187" }, processors: 8, memory: "64gb", storage: "128gb", version: "7.16.0" }
```
>processors should likely be 1, otherwise it will not pass node.processors validation?

In this case we're not validating the `node.processors` setting, so it should be fine. There's a test that checks the setting override `Test some settings can be overridden`.

> version going backwards seems like something we should reject up front?

That sounds sensible, I'm not sure if there's a way to inject the current version in these tests instead of a hardcoded value so the test don't fail after we move to another version.

> Can we also tear down the desired nodes to avoid them having an effect on follow-on tests?

We already do that in the setup section, did you mean something else?

henningandersen replied:
Yeah, just want it in a teardown: section instead.

fcofdez committed:
Add processors positive check

fcofdez committed:
Add new history id to response

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Fix docs

fcofdez committed:
Add teardown section

fcofdez committed:
DRY code a bit

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/DesiredNodesMetadataSerializationTests.java:
```diff
@@ -0,0 +1,74 @@
+        }
+        DesiredNodes latestDesiredNodes = base.getLatestDesiredNodes();
+        return new DesiredNodesMetadata(
+            new DesiredNodes(latestDesiredNodes.historyID(), latestDesiredNodes.version() + 1, latestDesiredNodes.nodes())
```
Can we make it also mutate the desired nodes?

henningandersen reviewed (commented):
A few more comments to consider.

fcofdez committed:
Populate version from cluster

fcofdez committed:
Rename node version to node_version

fcofdez committed:
Adjust desired nodes version type

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Fix doc tests

fcofdez committed:
Fallback to node.name on node.external_id

fcofdez committed:
Add version barrier

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Improve docs

fcofdez committed:
Fix docs

fcofdez commented:
Thanks for all the feedback, I think this is ready for another review round. I've addressed all the review comments, even though there are a couple of minor points that we need to clarify:
- How to override settings that have ranges defined using environmental values, `node.processors`
- How to report setting validation errors, currently we report the node position in the `nodes` list

@henningandersen @sethmlarson

henningandersen commented:
@elasticmachine run elasticsearch-ci/docs

henningandersen commented:
[Docs preview](https://elasticsearch_82975.docs-preview.app.elstc.co/diff)

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:
```diff
@@ -0,0 +1,124 @@
+
+        final Settings updatedSettings = unknownSettings.isEmpty() ? settings : updatedSettingsBuilder.build();
+        final ClusterSettings updatedClusterSettings = hasOverriddenSettings
+            ? new ClusterSettings(updatedSettings, settingSet)
```
I think we cannot really assume that this is safe. A couple of problems:

1. Assume an enum and the newer version added a new option. Using that would fail on current version. Similarly for any setting that has expanded set of legal values.
2. Assume a new setting is added that serves as the default for another setting and the orchestrator uses the new setting on a newer version rather than the original. This could in principle ruin validation.

We could verify the settings still and produce some sort of warning (I assume it will still be rare). But for now I would prefer to simply skip validation when the version is newer than current.

henningandersen commented on server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:
```diff
@@ -0,0 +1,124 @@
+            final DesiredNode node = nodes.get(i);
+            try {
+                validate(node);
+            } catch (RuntimeException e) {
```
Can we have other than `IllegalArgumentException` here? If not, I prefer to narrow it.

henningandersen reviewed (commented):
A few more comments. I believe I addressed the two questions inline too.

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportDeleteDesiredNodesAction.java:
```diff
@@ -0,0 +1,70 @@
+        ClusterState state,
+        ActionListener<AcknowledgedResponse> listener
+    ) throws Exception {
+        clusterService.submitStateUpdateTask("delete-desired-nodes", new AckedClusterStateUpdateTask(Priority.HIGH, request, listener) {
```
I think these don't need to be _acked_ tasks, we don't care whether the update is applied on all nodes or not since it will only be used on the master. It's enough that the state update is committed.

DaveCTurner replied:
> I think these don't need to be acked tasks, we don't care whether the update is applied on all nodes or not since it will only be used on the master. It's enough that the state update is committed.

I'd rather we didn't merge without addressing this comment.

fcofdez replied:
I addressed that comment in https://github.com/elastic/elasticsearch/pull/82975/commits/f558888a3a47936ad34d316fade0443a814005a9#diff-1570a3662da76672be15884ead7cab913541c1fc4176c36284166a2f3d7c34a6L79 unless I'm missing something.

DaveCTurner replied:
The `delete-desired-nodes` task is still an `AckedClusterStateUpdateTask` in beeccc077fb949c232fc4ce32e9954792db6e0b6  (the latest commit as I write this)

fcofdez replied:
Oh right 🤦, sorry about that.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    private void validate(DesiredNode node) {
+        if (node.version().before(Version.CURRENT)) {
```
Should this be `Version.V_8_1_0`?

henningandersen replied:
This was intentional. The idea is that a cluster should not be able to go backwards, thus the next desired nodes should always be newer than current.

DaveCTurner reviewed (commented):
I'm still reading but left a couple of comments from my first pass.

fcofdez committed:
Address review comments

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Docs

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Fix serialization tests

fcofdez committed:
Small fix

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportUpdateDesiredNodesAction.java:
```diff
@@ -0,0 +1,141 @@
+
+            clusterService.submitStateUpdateTask(
+                "update-desired-nodes",
+                new ClusterStateUpdateTask(Priority.HIGH, request.masterNodeTimeout()) {
```
I would think we should use `URGENT` here instead, but maybe @DaveCTurner has an opinion on that?

DaveCTurner replied:
I'd be ok with `URGENT` although I'd be happier about it if these things were batched, just in case the orchestrator goes haywire.

fcofdez replied:
I'll update the PR batching them

henningandersen commented on server/src/main/java/org/elasticsearch/action/admin/cluster/desirednodes/UpdateDesiredNodesRequest.java:
```diff
@@ -0,0 +1,114 @@
+        }
+
+        if (nodes.isEmpty()) {
+            validationException = ValidateActions.addValidationError("nodes must contain at least one node", validationException);
```
In a follow-up we should probably also verify that there is at least one master eligible node.

henningandersen commented on server/src/internalClusterTest/java/org/elasticsearch/action/admin/cluster/desirednodes/TransportDesiredNodesActionsIT.java:
```diff
@@ -166,7 +172,7 @@ public void testUnknownSettingsAreAllowedInFutureVersions() {
     }
 
-    public void testSomeSettingsCanBeOverridden() {
+    public void testNodeProcessorsGetValidatedWithDesiredNodeProcessors() {
```
I think this verifies what the method name says by not throwing when setting the desired nodes? Perhaps add a comment if so.

fcofdez commented on docs/reference/cluster/delete-desired-nodes.asciidoc:
```diff
@@ -0,0 +1,68 @@
+            "processors" : 8,
+            "memory" : "58gb",
+            "storage" : "1700gb",
+            "node_version" : "8.1.0"
```
I think there isn't a way to fetch this value programatically and populate it, I think we should upgrade it as new versions come out. Maybe it's better to skip the doc tests?

henningandersen replied:
I think it is ok to leave that for a follow-up at least.

henningandersen reviewed (commented):
Sorry for dumping yet another incomplete review, will get to the rest of it shortly.

fcofdez committed:
Review comments

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Check that at least one master node is present

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Fix docs...

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

henningandersen reviewed (approved):
LGTM. Thanks Francisco

fcofdez committed:
Batch tasks

fcofdez committed:
Merge remote-tracking branch 'origin/master' into desired-nodes-api

fcofdez committed:
Fix docs again...

DaveCTurner reviewed (approved):
Thanks @fcofdez LGTM

fcofdez commented:
Thanks all for the reviews!
